autonat/autonat-v2.md

```diff
@@ -0,0 +1,137 @@
+# AutonatV2: spec
+
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+|-----------------|--------------------------|--------|-----------------|
+| 1A              | Working Draft            | Active | r0, 2023-04-11  |
+
+Authors: [@sukunrt], [@marten-seemann]
+
+Interest Group: [@marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@sukunrt]: https://github.com/sukunrt
+
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is
+publicly reachable. Knowing its NAT status is essential for the node to be
+well-behaved in the network: A node that's behind a doesn't need to
+advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+`autonat v1` protocol was designed for a node to determine whether the node
+was publically rechable. However, it could not be tweaked to obtain more 
+granular information about which of the nodes addresses were publically 
+reachable. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they
+can test individual addresses discovered by different sources like identify, 
+upnp mappings, circuit addresses etc for reachability.
+
+
+## AutoNAT V2 Protocol
+
+The AutoNAT V2 Protocol uses the Protocol ID `/libp2p/autonat/2.0.0`. The node
+wishing to determine reachability of a given address sends a `DialRequest` 
+message. The `DialRequest` message contains a single multiaddress and a nonce 
+of 8 random bytes. Upon receiving  this message, the peer dials this address 
+and sends the nonce in the `Dial` message. The peer MUST NOT dial any address 
+other than the address in the DialRequest. The initiator SHOULD check that the 
+nonce received in the `Dial` message is the same as the nonce the initiator 
+sent in the `DialRequest` message. If the nonce received in the `Dial` message 
+is different the initiator should discard this `DialResponse`
+
+If the dial fails, the receiver sends a `DialResponse` message with the
+`ResponseStatus` `E_DIAL_ERROR`. If the dial completes successfully, it 
+sends a `DialResponse` with the `ResponseStatus` `OK`. Nodes MUST NOT send a 
+`E_DIAL_ERROR` status when they didn't attempt a dial. 
+
```

Comment:
We need some text that talks about refusal of dials.

There's two (main) reasons why a peer might refuse a dial request:
1. The peer might apply some kind of rate limiting / blacklisting.
2. Or it might not be able to parse the multiaddr, or not be able to dial the transport.

I'm wondering if it makes sense to split these two reason into separate response codes, as this allows the requester to adjust their response:
1. If we're rate limited / blacklisted, retrying with a different address has little potential for success.
2. If the peer just can't dial the respective multiaddr, we could just open a new stream and request dialing of another multiaddr.

We could also think of adding a list of supported transport multicodecs to that response message.

What do you think?

Reply:
I think separating them is a nice idea. Providing the list of supported transports might make the protocol more complex than we need. I'll change this to include the list of transports, we can then see if it is looking too complex.

Reply:
have separated response codes for these two cases. 

continuing the discussion for priority ordered address list here https://github.com/libp2p/specs/issues/539

Reply:
opened a PR for the rest of the changes. We can continue the discussion there.

Reply:
https://github.com/libp2p/specs/pull/541